File: examples/ch04/map.c
/** snippet uppercase */
#include <ctype.h>

char *uppercase(const char *in)
{
    char *out = strdup(in);
    
    if (out != NULL) {
	for (size_t i = 0; out[i] != '\0'; i++) {
	    out[i] = toupper(out[i]);
	}
    }

    return out;
}
/** /snippet uppercase */

/** snippet square */
void square(double *out, const double *in, size_t length)
{
    for (size_t i = 0; i < length; i++) {
	out[i] = in[i] * in[i];
    }
}
/** /snippet square */

File: examples/ch04/Map.hs
{-- snippet upperCase --}
import Data.Char (toUpper)

upperCase :: String -> String

upperCase (x:xs) = toUpper x : upperCase xs
upperCase []     = []
{-- /snippet upperCase --}

{-- snippet square --}
square :: [Double] -> [Double]

square (x:xs) = x*x : square xs
square []     = []
{-- /snippet square --}

{-- snippet map2 --}
square2 xs = map squareOne xs
    where squareOne x = x * x

upperCase2 xs = map toUpper xs
{-- /snippet map2 --}

{-- snippet myMap --}
myMap :: (a -> b) -> [a] -> [b]

myMap f (x:xs) = f x : myMap f xs
myMap _ _      = []
{-- /snippet myMap --}
